File: sdks/python/clicknest.py
"""ClickNest Python SDK for server-side analytics."""

import atexit
import json
import threading
import time
import urllib.request
from typing import Any, Optional


class ClickNest:
    """ClickNest analytics client for Python backends."""

    def __init__(
        self,
        api_key: str,
        host: str,
        flush_interval: float = 10.0,
        max_batch_size: int = 50,
    ):
        self.api_key = api_key
        self.host = host.rstrip("/")
        self.flush_interval = flush_interval
        self.max_batch_size = max_batch_size
        self._queue: list[dict] = []
        self._lock = threading.Lock()
        self._running = True

        self._timer = threading.Thread(target=self._flush_loop, daemon=True)
        self._timer.start()
        atexit.register(self.shutdown)

    def capture(
        self,
        event: str,
        distinct_id: Optional[str] = None,
        properties: Optional[dict[str, Any]] = None,
    ) -> None:
        """Track an event."""
        payload = {
            "event_type": "custom",
            "url": event,
            "url_path": event,
            "timestamp": int(time.time() * 1000),
            "properties": {**(properties or {}), "$event_name": event},
        }
        with self._lock:
            self._queue.append(
                {
                    "event": payload,
                    "distinct_id": distinct_id or "",
                }
            )
            if len(self._queue) >= self.max_batch_size:
                self._flush()

    def identify(self, distinct_id: str, previous_id: Optional[str] = None) -> None:
        """Link an anonymous ID to an identified user."""
        if previous_id:
            self.capture(
                "$identify",
                distinct_id=distinct_id,
                properties={"previous_id": previous_id},
            )

    def flush(self) -> None:
        """Force flush all queued events."""
        with self._lock:
            self._flush()

    def shutdown(self) -> None:
        """Flush remaining events and stop the background thread."""
        self._running = False
        self.flush()

    def _flush(self) -> None:
        if not self._queue:
            return

        # Group events by distinct_id for batching.
        batches: dict[str, list[dict]] = {}
        for item in self._queue:
            did = item["distinct_id"]
            batches.setdefault(did, []).append(item["event"])
        self._queue = []

        for did, events in batches.items():
            body = json.dumps(
                {
                    "events": events,
                    "session_id": f"server-{int(time.time())}",
                    "distinct_id": did,
                }
            ).encode()

            req = urllib.request.Request(
                f"{self.host}/api/v1/events",
                data=body,
                headers={
                    "Content-Type": "application/json",
                    "X-API-Key": self.api_key,
                },
            )
            try:
                urllib.request.urlopen(req, timeout=5)
            except Exception:
                pass  # Best effort — don't crash the app.

    def _flush_loop(self) -> None:
        while self._running:
            time.sleep(self.flush_interval)
            with self._lock:
                self._flush()

File: sdks/ruby/clicknest.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

# ClickNest Ruby SDK for server-side analytics.
module ClickNest
  class Client
    # @param api_key [String]
    # @param host [String]
    # @param flush_interval [Integer] seconds between flushes (default 10)
    # @param max_batch_size [Integer] max events before auto-flush (default 50)
    def initialize(api_key:, host:, flush_interval: 10, max_batch_size: 50)
      @api_key = api_key
      @host = host.chomp("/")
      @max_batch_size = max_batch_size
      @queue = []
      @mutex = Mutex.new

      @running = true
      @thread = Thread.new do
        while @running
          sleep flush_interval
          flush
        end
      end

      at_exit { shutdown }
    end

    # Track an event.
    # @param event [String] event name
    # @param distinct_id [String, nil]
    # @param properties [Hash, nil]
    def capture(event, distinct_id: nil, properties: {})
      item = {
        event: {
          event_type: "custom",
          url: event,
          url_path: event,
          timestamp: (Time.now.to_f * 1000).to_i,
          properties: (properties || {}).merge("$event_name" => event),
        },
        distinct_id: distinct_id || "",
      }

      @mutex.synchronize do
        @queue << item
        send_batch if @queue.size >= @max_batch_size
      end
    end

    # Link an anonymous ID to an identified user.
    # @param distinct_id [String]
    # @param previous_id [String, nil]
    def identify(distinct_id, previous_id: nil)
      return unless previous_id

      capture("$identify", distinct_id: distinct_id, properties: { "previous_id" => previous_id })
    end

    # Force flush all queued events.
    def flush
      @mutex.synchronize { send_batch }
    end

    # Flush and stop the background thread.
    def shutdown
      @running = false
      flush
    end

    private

    def send_batch
      return if @queue.empty?

      items = @queue.dup
      @queue.clear

      # Group by distinct_id.
      batches = items.group_by { |i| i[:distinct_id] }

      batches.each do |did, group|
        body = {
          events: group.map { |i| i[:event] },
          session_id: "server-#{Time.now.to_i}",
          distinct_id: did,
        }.to_json

        uri = URI("#{@host}/api/v1/events")
        req = Net::HTTP::Post.new(uri)
        req["Content-Type"] = "application/json"
        req["X-API-Key"] = @api_key
        req.body = body

        Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == "https", open_timeout: 5, read_timeout: 5) do |http|
          http.request(req)
        end
      rescue StandardError
        # Best effort — don't crash the app.
      end
    end
  end
end
